java/ql/src/semmle/code/java/frameworks/ratpack/RatpackExec.qll

```diff
@@ -0,0 +1,160 @@
+import java
+private import semmle.code.java.dataflow.DataFlow
+private import semmle.code.java.dataflow.FlowSteps
+
+/** A reference type that extends a parameterization the Promise type. */
+private class RatpackPromise extends RefType {
+  RatpackPromise() {
+    getSourceDeclaration().getASourceSupertype*().hasQualifiedName("ratpack.exec", "Promise")
+  }
+}
+
+private class RatpackPromiseValueMethod extends Method, TaintPreservingCallable {
+  RatpackPromiseValueMethod() { isStatic() and hasName("value") }
+
+  override predicate returnsTaintFrom(int arg) { arg = 0 }
+}
+
+abstract private class FluentLambdaMethod extends Method {
+  /**
+   * Holds if this lambda consumes taint from the quaifier when `lambdaArg`
+   * for `methodArg` is tainted.
+   * Eg. `tainted.map(stillTainted -> ..)`
+   */
+  abstract predicate consumesTaint(int methodArg, int lambdaArg);
+
+  /**
+   * Holds if the lambda passed at the given `arg` position produces taint
+   * that taints the result of this method.
+   * Eg. `var tainted = CompletableFuture.supplyAsync(() -> taint());`
+   */
+  predicate doesReturnTaint(int arg) { none() }
+}
+
+private class RatpackPromiseProviderMethod extends Method, FluentLambdaMethod {
+  RatpackPromiseProviderMethod() { isStatic() and hasName(["flatten", "sync"]) }
+
+  override predicate consumesTaint(int methodArg, int lambdaArg) { none() }
+
+  override predicate doesReturnTaint(int arg) { arg = 0 }
+}
+
+abstract private class SimpleFluentLambdaMethod extends FluentLambdaMethod {
+  override predicate consumesTaint(int methodArg, int lambdaArg) {
+    methodArg = 0 and consumesTaint(lambdaArg)
+  }
+
+  /**
+   * Holds if this lambda consumes taint from the quaifier when `arg` is tainted.
+   * Eg. `tainted.map(stillTainted -> ..)`
```

Comment:
Please clarify. What does "this lambda" refer to? It's not referring to `this` as far as I can see.